Copter: fix unconstrained avoidance backup speed

IamPete1 opened the pull request:
This fix is the result of a log where home changed altitude resulting in a huge breach of the max altitude fence. There was no fence action configured, however avoidance was enabled resulting in the vehicle trying to back away from the max altitude fence at 15 m/s. The configured climb rate limits were +-2.5 m/s. 

![image](https://github.com/ArduPilot/ardupilot/assets/33176108/a044929a-98fe-4b94-b393-c23dde79de3a)

Firstly this fixes the avoidance to constrain the velocity to `AVOID_BACKUP_SPD` parameter.

Secondly this constrains Z backup to half the set pos controller speeds. This helps keep the speeds in balance since the the avoid backup speed is probably set with XY speed in mind and could be too large for Z. 

I have not tested this yet. We might want a minimal fix to just apply the backup speed in avoid for a backport.

rishabsingh3003 commented on ArduCopter/mode.cpp:
```diff
@@ -948,7 +948,13 @@ float Mode::get_pilot_desired_throttle() const
+        // Ensure back up action keeps to the climbrate limits of the mode
+        // Apply 50% knock down to ensure backup is not too exciting
+        const float knock_down = 0.5;
+        return constrain_float(target_rate, pos_control->get_max_speed_down_cms() * knock_down, pos_control->get_max_speed_up_cms() * knock_down);
```
Do we really need this additional constraint? Is this going to change how "backing up" feels to the users already using it? 

I ask this since we are already constraining the back up in AC_Avoid (added in this PR)

IamPete1 replied:
The thinking here is that there is only one back up speed in AC_Avoid. So a user might set that faster with horizontal speed in mind, say 5m/s. That would be fine in XY for most vehicles but could be too much for descending. Ultimately the vehicle should never move faster than the limits set for the given mode. If I could only have one constrain it would be this one. I can see a argument for not using the 50% knock down tho....

rishabsingh3003 reviewed (approved):
Looks good to me! One minor comment. Thanks so much for fixing this.

IamPete1 commented:
I have reworked to take a pos control pointer from copter which allows the constrain to pos control limits to be done inside AC_Avoid resulting in a much smaller change. 

AC_Avoid is also used by rover which will not set the pointer so will not try and use pos control.

A alternative method might be to add a pos control singleton.

rmackay9 commented:
I'm slightly surprised that the vertical speed limits weren't being applied in AC_PosControl but perhaps that's because we are not using "input shaping"

IamPete1 commented:
> I'm slightly surprised that the vertical speed limits weren't being applied in AC_PosControl but perhaps that's because we are not using "input shaping"

The pos controller does not apply any limits, its on the caller to restrict what they ask for. I do think it should (https://github.com/ArduPilot/ardupilot/pull/24100) this is a case where having hard pos control limits would have been a good idea.

peterbarker commented on Tools/autotest/arducopter.py:
```diff
@@ -1603,6 +1603,58 @@ def MaxAltFence(self):
+
+        # Avoid should prevent the vehicle flying past the fence, so the altitude wait should timeouts
+        try:
+            self.wait_altitude(140, 150, timeout=90, relative=True)
```
```suggestion
            self.wait_altitude(140, 15000, timeout=90, relative=True)
```

removes the race condition

You could also simplify this to simply ensure you remain just below the fence instead, removing the try/except/except block

IamPete1 replied:
Sorry, I don't understand. Increasing the upper bound would mean it always times out? But the test wants it to time out?

tridge replied:
peter is happy with it

rmackay9 commented:
We discussed this on the call and agreed to use a new AVOID_BACKZ_SPD parameter.

IamPete1 commented:
Reworked to separate the speed limit into horizontal and vertical components. I also found I had omitted to apply the Z limit in the 3D case which I have now fixed.

rmackay9 reviewed (approved):
LGTM, thanks!

rmackay9 commented:
@peterbarker, I think we can merge this unless you're comment above re the autotest is a blocker... so up to you.
